src/pp/if_parser.rs

```diff
@@ -0,0 +1,441 @@
+            PreprocessorError::UnexpectedEndOfInput,
+            self.location,
+        )))
+    }
+
+    fn expect_peek(&mut self) -> Step<Token> {
+        let val = self.peek()?;
+        val.ok_or(StepExit::Error((
+            PreprocessorError::UnexpectedEndOfInput,
+            self.location,
+        )))
+    }
+
+    fn add_define(&mut self, name: &str, location: Location) -> Step<Option<Token>> {
+        Ok(if let Some(define) = self.defines.get(name) {
+            if self.macro_expansion.len() == self.max_recursion_depth {
+                return Err(StepExit::Error((
+                    PreprocessorError::RecursionLimitReached,
+                    location,
+                )));
+            }
+
+            assert!(!define.function_like, "TODO: Handle function like defines");
+
+            self.macro_expansion.push(define.tokens.iter().peekable());
+
+            None
+        } else {
+            Some(Token {
+                value: TokenValue::Integer(Integer {
+                    value: 0,
+                    signed: true,
+                    width: 64,
+                }),
+                location,
+            })
+        })
+    }
+
+    fn handle_defined(&mut self) -> Step<i64> {
+        let next = self.expect_raw_peek()?;
+
+        match next.value {
+            TokenValue::Ident(ref name) => {
+                self.expect_raw_next()?;
+
+                Ok(self.defines.get(name).is_some() as i64)
+            }
+            TokenValue::Punct(punct) => {
+                if let Punct::LeftParen = punct {
+                    self.expect_next()?;
+
+                    let name_token = self.expect_raw_next()?;
+                    let name = match name_token.value {
+                        TokenValue::Ident(name) => Ok(name),
+                        value => Err(StepExit::Error((
+                            PreprocessorError::UnexpectedToken(value),
+                            name_token.location,
+                        ))),
+                    }?;
+
+                    let close_brace = self.expect_next()?;
+
+                    match close_brace.value {
+                        TokenValue::Punct(Punct::RightParen) => {
+                            Ok(self.defines.get(&name).is_some() as i64)
+                        }
+                        value => Err(StepExit::Error((
+                            PreprocessorError::UnexpectedToken(value),
+                            close_brace.location,
+                        ))),
+                    }
+                } else {
+                    Err(StepExit::Error((
+                        PreprocessorError::UnexpectedToken(next.value),
+                        next.location,
+                    )))
+                }
+            }
+            value => Err(StepExit::Error((
+                PreprocessorError::UnexpectedToken(value),
+                next.location,
+            ))),
+        }
+    }
+
+    fn parse_atom(&mut self) -> Step<i64> {
```

Comment:
The rest of the library has extensive tests and tries to reach 100% coverage. Could we have tests for the parser as well? We could make the tests only deal with define-free expressions (and test the interaction of defines and other macro expansion in `pp_tests.rs`).

This split would also help test the i64 value of the expression more easily (instead of doing `#if (expr) != expected` `#error`)